packages/lesswrong/lib/ultraFeed/UltraFeedRepo.ts

```diff
@@ -0,0 +1,413 @@
+import { recombeeApi, recombeeRequestHelpers } from '../../server/recombee/client';
+import { HybridRecombeeConfiguration } from '../../lib/collections/users/recommendationSettings';
+import { aboutPostIdSetting } from '@/lib/instanceSettings';
+
+const COMMENT_LOOKBACK_WINDOW = '90 days';
+
+export interface ThreadStatistics {
+  commentCount: number;
+  maxKarma: number;
+  sumKarma: number;
+  sumKarmaSquared: number;
+  averageKarma: number;
+  averageTop3Comments: number;
+}
+
+class UltraFeedRepo extends AbstractRepo<"Comments"> {
+  constructor() {
+    super(Comments);
+  }
+
+  public async getUltraFeedPostThreads(
+    context: ResolverContext,
+    limit = 20,
+    servedPostIds: Set<string> = new Set()
+  ): Promise<FeedFullPost[]> {
+    
+
+    const { currentUser } = context;
+    const recombeeUser = recombeeRequestHelpers.getRecombeeUser(context);
+    let displayPosts: FeedFullPost[] = [];
+
+    if (!recombeeUser) {
+      // eslint-disable-next-line no-console
+      console.warn("UltraFeedRepo: No Recombee user found. Cannot fetch hybrid recommendations.");
+    } else {
+      const settings: HybridRecombeeConfiguration = {
+        hybridScenarios: { fixed: 'hacker-news', configurable: 'recombee-lesswrong-custom' },
+        excludedPostIds: Array.from(servedPostIds),
+        filterSettings: currentUser?.frontpageFilterSettings,
+      };
+
+      try {
+        const recommendedResults = await recombeeApi.getHybridRecommendationsForUser(
+          recombeeUser,
+          limit,
+          settings,
+          context
+        );
+
+        displayPosts = recommendedResults.map((item, idx): FeedFullPost | null => {
+           if (!item.post?._id) return null;
+
+
+           // Try to determine the scenario - using the same logic hierarchy as RecombeePostsList.tsx
+           let scenario: string | undefined = item.scenario;
+           
+           if (!scenario) {
+             // Try to detect special posts first, then try recommId
+             const aboutPostId = aboutPostIdSetting.get();
+             if (aboutPostId && item.post._id === aboutPostId && idx === 0) {
+               scenario = 'welcome-post';
+             } else if (item.curated) {
+               scenario = 'curated';
+             } else if (item.stickied || item.post.sticky) { 
+               scenario = 'stickied';
+             } else if (item.recommId) {
+               if (item.recommId.includes('forum-classic')) {
+                 scenario = 'hacker-news';
+               } else if (item.recommId.includes('recombee-lesswrong-custom')) {
+                 scenario = 'recombee-lesswrong-custom';
+               }
+             } else {
+               // Default fallback: Use empty sources array and warn
+               scenario = undefined; // Explicitly set scenario to undefined for clarity
+             }
+           }
+           
+           let sources: FeedItemSourceType[];
```

Comment:
why isn't this:
```
const sources: FeedItemSourceType[] = [];
if (scenario) {
  sources.push(scenario);
} else {
  console.warn(...);
}
```

(Though if the console.warn is still necessary I'm worried that something else is broken, so I'd prefer to fix that and then get rid of it entirely.)